jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Utf8Appendable.java

```diff
@@ -327,40 +331,159 @@ public boolean isUtf8SequenceComplete()
+            {
+                throw new RuntimeException(e);
+            }
+            _hasReplacements = true;
         }
+        _codep = 0;
+        _state = UTF8_ACCEPT;
     }
 
     /**
-     * Check the state of the Appendable to know if the current UTF-8 sequence is complete (or not)
+     * Clear out the {@link Appendable} of any content, resetting it to zero.
+     * <p>
+     *     This will not clear out any partial code points.
+     * </p>
      */
-    public void checkState()
+    public abstract void clear();
+
+    /**
+     * True if there was a replacement of a bad UTF-8 sequence with a UTF-8 replacement character.
+     *
+     * <p>
+     *     If the input already has a replacement character, this is not counted when using
+     *     this method.
+     * </p>
+     *
+     * @return true if there was a replacement of an appended character or byte with a replacement character.
+     */
+    public boolean hasReplacements()
     {
-        if (!isUtf8SequenceComplete())
+        return _hasReplacements;
+    }
+
+    /**
+     * Get the UTF-8 encoded String.
+     *
+     * <p>
+     *     This will get the String as it exists currently, without including any trailing incomplete UTF-8 sequences.
+     * </p>
+     * <p>
+     *     Use {@link #finish()} to complete any incomplete UTF-8 sequences before calling this method.
+     * </p>
+     *
+     * @param throwOnReplacement to throw if a replacement to an appended character has occurred
+     * @return the UTF-8 String
+     * @throws NotUtf8Exception if String has replacement characters due to bad UTF-8 sequences
+     * @see #finish()
+     */
+    public String getString(boolean throwOnReplacement)
+    {
+        if (throwOnReplacement && hasReplacements())
         {
-            if (_coderErrorAction == CodingErrorAction.REPLACE)
-            {
-                try
-                {
-                    _appendable.append(REPLACEMENT);
-                }
-                catch (IOException e)
-                {
-                    throw new RuntimeException(e);
-                }
-                _hasReplacements = true;
-            }
-            _codep = 0;
-            _state = UTF8_ACCEPT;
-            if (_coderErrorAction == CodingErrorAction.REPORT)
-                throw new NotUtf8Exception("incomplete UTF8 sequence");
+            throw new NotUtf8Exception("String has invalid UTF-8 sequences");
+        }
+        return _appendable.toString();
+    }
+
+    /**
+     * Get the String as it exists currently.
+     *
+     * <p>
+     *     This will get the String as it exists currently, without including any trailing incomplete UTF-8 sequences.
+     * </p>
+     * <p>
+     *     Use {@link #finish()} to complete any incomplete UTF-8 sequences before calling this method.
+     * </p>
+     *
+     * @return the String as it exists currently, without throwing an Exception
+     * @see #finish()
+     */
+    public String getString()
```

Comment:
I prefer to keep the methods as `toString()`, `toString(boolean)`, `takeString()` and `takeString(boolean)`.

I understand that many other classes have an informative `toString()` and use something like `asString()` to get the value.   However, this class is meant to be a drop in replacement to `StringBuilder` and `StringBuffer` and they use `toString()` to access the built value. 

You can see by the number of changed files in this PR how changing to `getString()` makes this a much larger change.

Reply:
`takeString()` is part of the `CharsetStringBuilder` interface that `Utf8Appendable` implements (and is used in many places).
It must throw a checked `CharacterCodingException` (which is the opposite of what we want).

The original `toString()` was changing the state of the internals of the class (to satisfy unfinished sequences), which broke debugging (it was impossible to debug a `Utf8Appendable` implementation while the debugger itself calls `.toString()` to show things, while the state internally changes), and if you enabled logging you broke the behavior of an active `Utf8Appendable` too for the same reason.   
That nuance, the changing of state during `toString()` was the cause of several other bugs as well.

There's 4 read access type demands on an `Utf8Appendable` implementation.

1. The entire String, with unfinished codepoints/sequences resolved. (optional boolean for throwing)
2. The entire String, without codepoints/sequences resolved. (optional boolean for throwing)
3. The String as it sits now, and then empty/clear the buffer, but not the unfinished codepoints/sequences. (optional boolean for throwing)
4. The `CharsetStringBuilder` in UTF-8 mode (always throws `CharacterCodingException`)

The most common use case is 1, which mimics the StringBuilder and StringBuffer behavior. (approx 75% of cases)
Use case 2 and 3 are very similar IMO.
Use case 2 is for when you want to access the content before it's finished. (few uses, and I think we can drop this as a use case in favor of just using other methods)
Use case 3 is the predominant WebSocket use case.
Use case 4 is used by server FormFields and IO ContentSource implementations.

For Use case 1, the behavior is like this ...
``` java
Utf8StringBuilder buffer = new Utf8StringBuilder();
buffer.append(bytes);
buffer.finish();
String result = buffer.toString();
```

For Use case 2, the behavior is like this ...
``` java
Utf8StringBuilder buffer = new Utf8StringBuilder();
buffer.append(bytes);
String partialResults = buffer.toString();
buffer.append(bytes);
buffer.finish();
String results = buffer.toString();
```

For Use case 3, the behavior is like this ...
``` java
Utf8StringBuilder buffer = new Utf8StringBuilder();
buffer.append(bytes);
String result = buffer.toString();
buffer.clear();
```

For Use case 4, the behavior is ...
``` java
CharsetStringBuilder text = CharsetStringBuilder.forCharset(StandardCharsets.UTF_8); // uses Utf8StringBuilder
text.append(bytes);
String result = text.takeString();
```

Whatever solution we settle on we should not have `toString()` cause internal state changes, and still support the `CharsetStringBuilder` concepts somehow (maybe not all the way down in `Utf8Appendable`).

Reply:
Good point about debugging. Hmmm no easy solutions. Ok let's go with getString and takeString.

But why takeFinishedString?

Reply:
Is that instead of takePartialString?

I prefer having partial string methods and the getString ones will always be finished strings

Reply:
resolving so I can see current state clearly.